argus/action_manager/windows.py

```diff
@@ -115,32 +117,83 @@ def get_installation_script(self):
+            return False
+
+        LOG.info("Cloudbase-Init was successfully installed!")
+        return True
+
+    def cbinit_cleanup(self, cbinit_dir):
+        """Clean up Cloudbase-Init."""
+        LOG.info("Cleaning up Cloudbase-Init ...")
+
+        self.rmdir(ntpath.dirname(cbinit_dir))
+
     def install_cbinit(self):
-        """Run the installation script for CloudbaseInit."""
-        LOG.debug("Installing Cloudbase-Init ...")
+        """Run the installation script for Cloudbase-Init."""
+        LOG.debug("Trying to install Cloudbase-Init ...")
 
         installer = "CloudbaseInitSetup_{build}_{arch}.msi".format(
             build=self._conf.argus.build,
             arch=self._conf.argus.arch
         )
         # TODO(cpopa): the service type is specific to each scenario,
         # find a way to pass it
-        LOG.info("Run the downloaded installation script "
-                 "using the installer %r.", installer)
 
-        parameters = '-installer {}'.format(installer)
-        try:
-            self.execute_powershell_resource_script(
-                resource_location='windows/installCBinit.ps1',
-                parameters=parameters)
-        except exceptions.ArgusError:
-            # This can happen for multiple reasons,
-            # but one of them is the fact that the installer
-            # can't be installed through WinRM on some OSes
-            # for whatever reason. In this case, we're falling back
-            # to use a scheduled task.
-            LOG.debug("Cannot install, deploying using a scheduled task.")
-            self._deploy_using_scheduled_task(installer)
+        for _ in range(util.RETRY_COUNT):
+            LOG.info("Running the downloaded installation script "
+                     "using the installer %r.", installer)
+
+            parameters = '-installer {}'.format(installer)
+            try:
+                self.execute_powershell_resource_script(
+                    resource_location='windows/installCBinit.ps1',
+                    parameters=parameters)
+            except exceptions.ArgusError:
+                # This can happen for multiple reasons,
+                # but one of them is the fact that the installer
+                # can't be installed through WinRM on some OSes
+                # for whatever reason. In this case, we're falling back
+                # to use a scheduled task.
+                LOG.debug("Could not install Cloudbase-Init!")
+
+                LOG.info("Deploying Cloudbase-Init using a scheduled task ...")
+                try:
+                    self._deploy_using_scheduled_task(installer)
+                except exceptions.ArgusError:
+                    LOG.debug("Could not install Cloudbase-Init "
+                              "using a scheduled task!")
+
+                    LOG.debug("Retrying ...")
+                    continue
+
+            python_dir = introspection.get_python_dir(self._execute)
```

Comment:
What of there is no `python_dir` ? If we fail installation we might not have Cloudbase-Init directory, and if we don't have one we won't have any python directory.This function `get_python_dir` uses `get_cbinit_dir` that raises an Exception if no Cloudbase-Init directory could be found.

https://github.com/cloudbase/cloudbase-init-ci/blob/master/argus/introspection/cloud/windows.py#L155